src/timezones/interpret.jl

```diff
@@ -0,0 +1,193 @@
+# TimeZone concepts used to disambiguate context of DateTimes
+# abstract UTC <: TimeZone  # Defined in Base.Dates
+abstract Local <: TimeZone
+
+function transition_range(local_dt::DateTime, tz::VariableTimeZone, ::Type{Local})
+    transitions = tz.transitions
+
+    # Determine the earliest and latest possible UTC DateTime
+    # that this local DateTime could be.
+    # TODO: Alternatively we should only look at the range of offsets available within
+    # this TimeZone.
+    earliest = local_dt + MIN_OFFSET
+    latest = local_dt + MAX_OFFSET
+
+    # Determine the earliest transition the local DateTime could
+    # occur within.
+    start = searchsortedlast(
+        transitions, earliest,
+        by=el -> isa(el, Transition) ? el.utc_datetime : el,
+    )
+    start = max(start, 1)
+    finish = length(transitions)
+    for i in start:finish
+        if transitions[i].utc_datetime > latest
+            finish = i - 1
+            break
+        end
+    end
+
+    return start:finish
+end
+
+function transition_range(utc_dt::DateTime, tz::VariableTimeZone, ::Type{UTC})
+    index = searchsortedlast(
+        tz.transitions, utc_dt,
+        by=el -> isa(el, Transition) ? el.utc_datetime : el,
+    )
+    index = max(index, 1)
+    return index:index
+end
+
+"""
+    transition_range(dt::DateTime, tz::VariableTimeZone, context::Type{Union{Local,UTC}}) -> UnitRange
+
+Finds the indexes of the `tz` transitions which may be applicable for the `dt`. The given
+DateTime is expected to be local to the time zone or in UTC as specified by `context`. Note
+that UTC context will always return a range of length one.
+"""
+transition_range(dt::DateTime, tz::VariableTimeZone, context::Type{Union{Local,UTC}})
```

Comment:
It really looks like a function call. I'd prefer the arguments go unnamed a la 
```julia
transition_range(::DateTime, ::VariableTimeZone, ::Type{Union{Local,UTC}})
```
which is also valid syntax.